cctp_primitives/src/bit_vector/merkle_tree.rs

```diff
@@ -0,0 +1,163 @@
+//! # Merkle Tree
+//!
+//! `merkle_tree` exposes functions to compute a bit vector Merkle tree.
+
+use super::compression;
+
+use algebra::{
+    fields::tweedle::Fr as TweedleFr, ToConstraintField
+};
+
+use primitives::{
+    crh::poseidon::parameters::tweedle::{TweedleFrPoseidonHash, TweedleFrBatchPoseidonHash},
+    merkle_tree::field_based_mht::{
+        FieldBasedMerkleTreeParameters, FieldBasedMerkleTreePrecomputedEmptyConstants,
+        BatchFieldBasedMerkleTreeParameters, FieldBasedOptimizedMHT, FieldBasedMerkleTree,
+        parameters::tweedle_fr::TWEEDLE_MHT_POSEIDON_PARAMETERS
+    }
+};
+
+use bit_vec::BitVec;
+
+#[derive(Clone, Debug)]
+struct TweedleFieldBasedMerkleTreeParams;
+impl FieldBasedMerkleTreeParameters for TweedleFieldBasedMerkleTreeParams {
+    type Data = TweedleFr;
+    type H = TweedleFrPoseidonHash;
+    const MERKLE_ARITY: usize = 2;
+    const EMPTY_HASH_CST: Option<FieldBasedMerkleTreePrecomputedEmptyConstants<'static, Self::H>> = Some(TWEEDLE_MHT_POSEIDON_PARAMETERS);
+}
+
+impl BatchFieldBasedMerkleTreeParameters for TweedleFieldBasedMerkleTreeParams {
+    type BH = TweedleFrBatchPoseidonHash;
+}
+
+type TweedlePoseidonMHT = FieldBasedOptimizedMHT<TweedleFieldBasedMerkleTreeParams>;
+
+type Error = Box<dyn std::error::Error>;
+
+const MERKLE_TREE_HEIGHT: usize = 12;
+
+/// Computes the root hash of the Merkle tree created as a representation
+/// of `uncompressed_bit_vector`.
+///
+/// The input byte array is expected to have a Big Endian bit order, for example:
+/// Bit Array [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0] <=> Byte Array [1, 2]
+/// 
+/// # Examples
+///
+/// ```
+/// use cctp_primitives::bit_vector::merkle_tree::*;
+///
+/// let bit_vector: Vec<u8> = (0..100).collect();
+/// let merkle_root = merkle_root_from_bytes(&bit_vector).unwrap();
+/// 
+/// ```
+pub fn merkle_root_from_bytes(uncompressed_bit_vector: &[u8]) -> Result<algebra::Fp256<algebra::fields::tweedle::FrParameters>, Error> {
+
+    let bv = BitVec::from_bytes(&uncompressed_bit_vector);
+    let bool_vector: Vec<bool> = bv.into_iter().map(|x| x).collect();
+
+    let num_leaves = 1 << MERKLE_TREE_HEIGHT;
```

Comment:
merkle tree size must be equal to bit vector size

Reply:
Fixed.